streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala

```diff
@@ -0,0 +1,100 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.scheduler.rate
+
+/**
+ * Implements a proportional-integral-derivative (PID) controller which acts on
+ * the speed of ingestion of elements into Spark Streaming. A PID controller works
+ * by calculating an '''error''' between a measured output and a desired value. In the
+ * case of Spark Streaming the error is the difference between the measured processing
+ * rate (number of elements/processing delay) and the previous rate.
+ *
+ * @see https://en.wikipedia.org/wiki/PID_controller
+ *
+ * @param batchDurationMillis the batch duration, in milliseconds
+ * @param proportional how much the correction should depend on the current
+ *        error. This term usually provides the bulk of correction. A value too large would
+ *        make the controller overshoot the setpoint, while a small value would make the
+ *        controller too insensitive. The default value is -1.
+ * @param integral how much the correction should depend on the accumulation
+ *        of past errors. This term accelerates the movement towards the setpoint, but a large
+ *        value may lead to overshooting. The default value is -0.2.
+ * @param derivative how much the correction should depend on a prediction
+ *        of future errors, based on current rate of change. This term is not used very often,
+ *        as it impacts stability of the system. The default value is 0.
+ */
+private[streaming] class PIDRateEstimator(
+    batchIntervalMillis: Long,
+    proportional: Double = -1D,
+    integral: Double = -.2D,
+    derivative: Double = 0D)
+  extends RateEstimator {
+
+  private var firstRun: Boolean = true
+  private var latestTime: Long = -1L
+  private var latestRate: Double = -1D
+  private var latestError: Double = -1L
+
+  require(
+    batchIntervalMillis > 0,
+    s"Specified batch interval $batchIntervalMillis in PIDRateEstimator is invalid.")
+
+  def compute(time: Long, // in milliseconds
+      elements: Long,
+      processingDelay: Long, // in milliseconds
+      schedulingDelay: Long // in milliseconds
+    ): Option[Double] = {
+
+    this.synchronized {
+      if (time > latestTime && processingDelay > 0 && batchIntervalMillis > 0) {
+
+        // in seconds, should be close to batchDuration
+        val delaySinceUpdate = (time - latestTime).toDouble / 1000
+
+        // in elements/second
+        val processingRate = elements.toDouble / processingDelay * 1000
+
+        // in elements/second
+        val error = latestRate - processingRate
+
+        // in elements/second
+        val sumError = schedulingDelay.toDouble * processingRate / batchIntervalMillis
```

Comment:
Carrying over conversation from previous thread that got lost due to rebase

>Its hard to understand what sumError mean in terms of the rates and all? Can you write down the physical interpretation of this sumError? And also make the name better accordingly? 
cc @huitseeker
@tdas
tdas added a note 14 hours ago
So I am trying to understand this. 
(scheduling delay / batch interval) = approx the number of batches the system is delayed. Lets call it numDelayedBatches.
Now you are multiplying numDelayedBatches X processingSpeed. So you are scaling the current processing rate with number of batches that are delayed. Right?

Reply:
> from the number of overflowing elements we can calculate the rate at which they would be cleared by dividing it by the batch interval

@dragos Why? I still don't understand why historical error = # overflowing elements / batch interval. The explanation that if # overflowing elements = 0, then historical error = 0 does not explain the 1 / batch interval coefficient.

I feel the hard part is: It is easy to reason about the accumulated error if error is an absolute number, for example, the number of overflowing elements, but what is an accumulated error of a rate?

Setting up 1 / batch interval as the coefficient may not be a big problem anyway, because at the end of the day it is nothing but a constant that you can tune the integral coefficient to compensate.

Reply:
@YikSanChan sorry for the delay, it's been a while since I didn't visit this code. I admit I am not sure I understand where you're coming from, and I don't want to simply repeat what I wrote above. So I'll ask a question or two myself first:

What prompted you to look into this code? Do you see bad behavior on certain loads?

> The explanation that if # overflowing elements = 0, then historical error = 0 does not explain the 1 / batch interval coefficient.

If the overflowing elements are 0, doesn't this imply the `schedulingDelay` is zero, so this whole term is 0 as well (instead of `1/batchInterval`)?

>I feel the hard part is: It is easy to reason about the accumulated error if error is an absolute number, for example, the number of overflowing elements, but what is an accumulated error of a rate?

If we agree that the accumulated error is what it is computed above as number of elements, then "all we need to do" is to transform it into a rate (elements/second). We need to do that since that's the quantity that this PID is controlling. By dividing it by the batch interval we get how many `elements/second` are overflowing, in other words, by how much is the existing rate "off", if we wanted to have no overflowing elements. Does this make sense to you?

Reply:
@dragos Thanks for responding!

> What prompted you to look into this code? Do you see bad behavior on certain loads?

I am not seeing bad behavior because I am not using spark streaming. The reason why I look into the code is I try to leverage the PID controller in an imaginary scenario I describe [here](https://robotics.stackexchange.com/questions/20248/how-to-determine-actuator-and-process-for-pid-and-how-to-intepret-the-output). I use Scala and Akka so I look for PID controller example in Scala on GitHub, which brings me here.

Your implementation gives me a nice picture of how to turn a "rule of thumb" (that's how I feel about PID) into an actual implementation that is well explained and understood (with clear physical/system interpretation) and runs well in production. Thank you :)

> If the overflowing elements are 0, doesn't this imply the schedulingDelay is zero, so this whole term is 0 as well (instead of 1/batchInterval)?

Yes, you are right. But what I am asking is: in the equation `historicalError = schedulingDelay * processingRate / batchInterval`, why is the coefficient `1 / batchInterval`, rather than `1 / 2`, `1 / 100`, `1 / sum(batchInteval)`, or whatever else?

> in other words, by how much is the existing rate "off", if we wanted to have no overflowing elements.

If I understand correctly, the overflowing elements are accumulated since batch 0 (provided we're at batch n). If we all agree with the statement, then shouldn't the rate be `1 /(n * batchInterval)` instead?